optimistix/_solver/bisection.py

```diff
@@ -18,6 +19,63 @@ class _BisectionState(eqx.Module, strict=True):
     error: Float[Array, ""]
 
 
+class _ExpansionCarry(eqx.Module, strict=True):
+    lower: Scalar
+    upper: Scalar
+
+
+def _interval_contains_root(
+    carry: _ExpansionCarry,
+    /,
+    *,
+    need_positive: Bool[Array, ""],
+    expand_upper: Bool[Array, ""],
+    fn: Fn[Scalar, Scalar, Aux],
+    args: Any,
+) -> Bool[Array, ""]:
+    new_boundary = jnp.where(expand_upper, carry.upper, carry.lower)
+    carry_val, _ = fn(new_boundary, args)
+    return need_positive ^ (carry_val > 0.0)
+
+
+def _expand_interval(
+    carry: _ExpansionCarry,
+    /,
+    *,
+    expand_upper: Bool[Array, ""],
+) -> _ExpansionCarry:
+    new_domain = 2.0 * (carry.upper - carry.lower)
```

Comment:
Why `2 * `? This means we triple the length of the interval each time?

Reply:
Sorry if this was confusing, but I shift lower and upper while doubling the interval.

For example:
lower = 1
upper = 4
f(lower) = -10
f(upper) = -5

So, we want to move right.  We know the root is not in [1, 4], and we want to double the interval length.  Thus:

new_lower = 4
new_upper = 2 * 3 + 4 = 10

What do you think?

Reply:
Ah, understood! LGTM :)